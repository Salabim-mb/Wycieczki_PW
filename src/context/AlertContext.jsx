import React, {useState} from 'react';
import alertStyle from 'constants/alertOptions';

export const AlertContext = React.createContext({
	open: false,
	message: "",
	showAlert: () => {}
});

export const AlertProvider = (props) => {
	const [open, setOpen] = useState(false);
	const [message, setMessage] = useState("");
	const [severity, setSeverity] = useState(alertStyle.ERROR);

	const data = {
		open,
		message,
		severity,
		changeVisibility: (newOpen) => setOpen(newOpen),
		changeMessage: (newMessage, newSeverity=alertStyle.ERROR) => {
			setOpen(true);
			setMessage(newMessage);
			setSeverity(newSeverity);
		},
		showAlert: (newMessage, newSeverity) => {
			data.changeMessage(newMessage, newSeverity);
			data.changeVisibility(true);
		},
	};

	return <AlertContext.Provider value={data} {...props} />
};