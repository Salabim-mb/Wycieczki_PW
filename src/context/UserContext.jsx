import React, { useState } from 'react';
import Cookies from 'universal-cookie';
import { decryptCookie, encryptCookie } from 'utils/aesCiphrator';

const cookie = new Cookies();

export const UserContext = React.createContext({
	token: undefined,
	data: undefined,
	type: undefined,
	login: () => {},
	logout: () => {}
});

export const UserProvider = (props) => {
	const [token, setToken] = useState(
		decryptCookie(cookie.get("token")) || undefined
	);
	const [data, setData] = useState(
		decryptCookie(cookie.get("data")) || undefined
	);
	const [type, setType] = useState(
		decryptCookie(cookie.get("type")) || undefined
	)

	const user = {
		token,
		data,
		type,
		login: (newToken, newData, newType) => {
			setToken(newToken);
			setData(newData);
			setType(newType);
			cookie.set("token", encryptCookie(newToken), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
			cookie.set("data", encryptCookie(newData), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
			cookie.set("type", encryptCookie(newType), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
		},
		logout: () => {
			cookie.remove("token");
			cookie.remove("data");
			cookie.remove("type");
			setData(undefined);
			setToken(undefined);
			setType(undefined);
		},
	};

	return <UserContext.Provider value={user} {...props} />;
};