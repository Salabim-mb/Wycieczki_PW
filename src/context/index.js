export * from './AlertContext';
export * from './UserContext';
