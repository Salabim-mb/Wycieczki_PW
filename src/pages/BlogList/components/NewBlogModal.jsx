import React, { useContext, useState } from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { Checkbox, FormControlLabel, TextField } from '@material-ui/core';
import PhotoCamera from '@material-ui/icons/PhotoCamera';
import toBase64 from 'utils/fileToBase64';
import {paths} from 'constants/paths';
import api from 'constants/api';
import {Redirect} from "react-router-dom";
import { AlertContext } from 'context/AlertContext';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
	root: {
		'& > *': {
			margin: theme.spacing(1),
		},
	},
	input: {
		display: 'none',
	},
	formInput: {
		margin: theme.spacing(2, 0)
	},
	dialogContent: {
		alignContent: "center"
	}
}));

const addBlogTile = async (token, data) => {
	const url = api.BLOG;
	const res = await fetch(url, {
		headers: {
			"Content-Type": "multipart/form-data"
		},
		method: "POST",
		body: (() => {
			const fd = new FormData();
			Object.keys(data).forEach((item) => {
				fd.append(item, data[item])
			});
			return fd;
		})()
	});

	if (res.status === 201) {
		const parsedRes = await res.json();
		return parsedRes;
	}
	throw new Error(await res.json());
}

const NewBlogModal = ({open, setOpen}) => {
	const theme = useTheme();
	const fullScreen = useMediaQuery(theme.breakpoints.down('sm'));
	const [title, setTitle] = useState("");
	const [photo, setPhoto] = useState(null);
	const [showTitle, setShowTitle] = useState(false);
	const [redirect, setRedirect] = useState({
		action: false,
		destination: null
	});
	const alertC = useContext(AlertContext);
	let tilePhoto = React.createRef();

	const classes = useStyles();

	const handleAddPhoto = async () => {
		const file = tilePhoto.current.files[0];
		const currentPhoto = {
			img: file,
			imgb64: await toBase64(file) || null,
			title: file?.name
		};
		setPhoto(currentPhoto);
		tilePhoto = React.createRef();
	};

	const handleClose = () => {
		setTitle("");
		setPhoto(null);
		setShowTitle(false);
		setOpen(false);
	};

	const handleSubmit = async (event) => {
		event.preventDefault();
		const form = event.currentTarget;
		if (form.reportValidity() === false) {
			if (photo === null || photo?.img === undefined) {
				alertC.showAlert("Brak zdjęcia!");
			}
			event.stopPropagation();
		} else {
			try {
				const {id} = await addBlogTile(null, {
					image: photo.img,
					title,
					show_title: showTitle
				});
				setRedirect({
					action: true,
					destination: id,
				});
				handleClose();
			} catch(ex) {
				alertC.showAlert(ex.message);
			}
		}

	};

	return (
		<>
			<Dialog
				fullScreen={fullScreen}
				maxWidth="lg"
				open={open}
				onClose={handleClose}
				scroll="paper"
				aria-labelledby="responsive-dialog-title"
			>
				<DialogTitle id="responsive-dialog-title">Dodaj nowy blog</DialogTitle>
				<form onSubmit={handleSubmit} noValidate>
					<DialogContent className={classes.dialogContent}>
						<TextField
							id="outlined-basic"
							label="Tytuł bloga"
							required
							name="title"
							variant="outlined"
							fullWidth
							value={title}
							onChange={(e) => setTitle(e.target.value)}
							className={classes.formInput}
						/>
						<input
							ref={(ref) => tilePhoto.current = ref}
							accept="image/*"
							className={classes.input}
							id="icon-button-file"
							required
							name="photo"
							type="file"
							onChange={handleAddPhoto}
						/>
						{/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
						<label htmlFor="icon-button-file">
							<Button variant="contained" color="primary" component="span" startIcon={<PhotoCamera />}>
								Dodaj zdjęcie kafelka
							</Button>
						</label>
						<TextField
							disabled
							variant="outlined"
							name="photo.title"
							value={photo?.title}
							placeholder="Wybierz plik przyciskiem powyżej."
							className={classes.formInput}
							fullWidth
						/>
						<FormControlLabel
							control={
								<Checkbox
									checked={showTitle}
									onChange={(e) => setShowTitle(e.target.checked)}
									name="showTitle"
									color="primary"
								/>
							}
							label="Tytuł widoczny na kafelku"
						/>
					</DialogContent>
					<DialogActions>
						<Button type="submit" variant="contained" color="primary" autoFocus>
							Zatwierdź
						</Button>
						<Button autoFocus onClick={handleClose} color="primary">
							Anuluj
						</Button>
					</DialogActions>
				</form>
			</Dialog>
			{redirect?.action && <Redirect to={paths.BLOG_CATEGORIES.redirect(redirect?.destination)} />}
		</>
	);
}

NewBlogModal.propTypes = {
	open: PropTypes.bool.isRequired,
	setOpen: PropTypes.func.isRequired
}

export default NewBlogModal;