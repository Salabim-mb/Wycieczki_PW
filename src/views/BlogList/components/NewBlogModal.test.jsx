import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { AlertProvider } from 'context/AlertContext';
import NewBlogModal from './NewBlogModal';

describe("NewBlogModal", () => {
	let open;
	let setOpen;
	// let fetchFail;
	
	// beforeAll(() => {
	// 	global.fetch = jest.fn().mockImplementation(() => new Promise((resolve) => {
	// 			if (fetchFail) {
	// 				resolve({status: 500, json: () => Promise.resolve({message: "not ok"})});
	// 			} else {
	// 				resolve({status: 201, json: () => Promise.resolve({id: 1})});
	// 			}
	// 		}))
	// });

	beforeEach(() => {
		open = true;
		setOpen = jest.fn();
		// fetchFail = false;
		jest.clearAllMocks();
	});
	
	it("should match snapshot", () => {
		const {getByRole} = render(
			<AlertProvider>
				<NewBlogModal open={open} setOpen={setOpen} />
			</AlertProvider>
		);

		expect(getByRole("dialog")).toMatchSnapshot();
	});

	it("should clear data on cancel", () => {
		const {getByLabelText, getByText} = render(
			<AlertProvider>
				<NewBlogModal open={open} setOpen={setOpen} />
			</AlertProvider>
		);

		fireEvent.change(getByLabelText("Tytuł bloga", {exact: false}), {
			target: {
				value: "qwe"
			}
		});
		fireEvent.click(getByText("Anuluj"));

		expect(getByLabelText("Tytuł bloga", {exact: false}).value).toBe("");
	});

	it("should call fetching method with correct data", () => {
		const {getByLabelText, getByText} = render(
			<AlertProvider>
				<NewBlogModal open={open} setOpen={setOpen} />
			</AlertProvider>
		);

		fireEvent.change(getByLabelText("Tytuł bloga", {exact: false}), {
			target: {
				value: "qwe"
			}
		});
		fireEvent.change(getByLabelText("Dodaj zdjęcie kafelka", {exact: false}), {
			target: {
				files: [
					new File([], "qwe.jpg"),
				]
			}
		});
		fireEvent.change(getByLabelText("Tytuł widoczny na kafelku", {exact: false}), {
			target: {
				checked: true
			}
		});

		fireEvent.click(getByText("Zatwierdź"))
	});

	it("should display photo name in input", async () => {
		const {getByLabelText, getByPlaceholderText} = render(
			<AlertProvider>
				<NewBlogModal open={open} setOpen={setOpen} />
			</AlertProvider>
		);
		fireEvent.change(getByLabelText("Dodaj zdjęcie kafelka", {exact: false}), {
			target: {
				files: [
					new File(["qweqwe", "qwe"], "qwe.jpg"),
				]
			}
		});
		const fileNameInput = getByPlaceholderText(
			"Wybierz plik przyciskiem powyżej",
			{exact: false}
		)
		await expect(fileNameInput.value).toBe("qwe.jpg");
	});
});