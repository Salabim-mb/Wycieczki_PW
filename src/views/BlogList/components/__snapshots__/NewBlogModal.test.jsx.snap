// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewBlogModal should match snapshot 1`] = `
<div
  aria-labelledby="responsive-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthLg MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root"
    id="responsive-dialog-title"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6"
    >
      Dodaj nowy blog
    </h2>
  </div>
  <form
    novalidate=""
  >
    <div
      class="MuiDialogContent-root makeStyles-dialogContent-4"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-formInput-3 MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
          for="outlined-basic"
          id="outlined-basic-label"
        >
          Tytuł bloga
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="outlined-basic"
            name="title"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Tytuł bloga
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <input
        accept="image/*"
        class="makeStyles-input-2"
        id="icon-button-file"
        name="photo"
        required=""
        type="file"
      />
      <label
        for="icon-button-file"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="3.2"
                />
                <path
                  d="M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
                />
              </svg>
            </span>
            Dodaj zdjęcie kafelka
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-formInput-3 MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
          data-shrink="true"
        >
          Nazwa pliku
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
            disabled=""
            name="photo.title"
            type="text"
            value="Brak."
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
            >
              <span>
                Nazwa pliku
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-12"
              data-indeterminate="false"
              name="showTitle"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Tytuł widoczny na kafelku
        </span>
      </label>
    </div>
    <div
      class="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-focusVisible Mui-focusVisible"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Zatwierdź
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
            />
          </span>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-focusVisible Mui-focusVisible"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Anuluj
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
            />
          </span>
        </span>
      </button>
    </div>
  </form>
</div>
`;
